### PR TITLE
Global variables for explicit types

### PR DESCRIPTION
**Pull request by vlazzarini**
Currently there is no syntax for declaring global variables for explicit types. This PR introduces this, following the discussion in #2003

Here is the main outline of the solution:

-  We would like to provide a global declaration format that parallels the one used for implicit types.
-  Global variables can be added anywhere in the code 
-  The annotation `@global` to a variable has the same effect as `g` to an implicit type. 
-  This is added to the variable name when the type is declared.

Here's a test code as an example:

```
myvar@global:i init 0
myarr@global:i[] fillarray 1,2,3

instr 1
print myvar
print myarr[0]
myarr[1] = 0
endin
```

### CODE REVIEW

**Comment by vlazzarini** [comment]
this is complete from my side.

**Comment by vlazzarini** [comment]
Fixed the issue in the develop code with const instr refs

File: Engine/csound_orc.lex [hunk 1]
```diff
@@ -77,7 +77,7 @@ int get_next_char(char *, int32_t, struct yyguts_t*);
 %option nounput
 
 IDENT           [a-zA-Z_][a-zA-Z0-9_]*
-TYPED_IDENTIFIER  [a-zA-Z_][a-zA-Z0-9_]*:[a-zA-Z_][a-zA-Z0-9_]*
+TYPED_IDENTIFIER  [a-zA-Z_][a-zA-Z0-9_@]*:[a-zA-Z_][a-zA-Z0-9_]*
```
Comment (kunstmusik) [inline Engine/csound_orc.lex]: I think this rule is off as it permits multiple @ symbols. I need to check but it should probably use a ? for optional @global, something like :

[a-zA-Z_][a-zA-Z0-9_]*(@global)?:[a-zA-Z_][a-zA-Z0-9_]*

**Comment by kunstmusik** [comment]
Wanted to inquire, as I thought I had floated this idea, but why do we need @global? Could we instead just have scoping rules such that if you declare a variable in instr 0 space it's automatically in the global var pool and accessible by any instrument? Compiling of instruments should be able to lookup the var from the local var space first, then fallback to global, and you shouldn't need to have @global added. 

This would follow scoping rules of something like Javascript. Local names could shadow global ones and that would maintain backwards compatibility.

File: Engine/csound_orc_semantics.c [hunk 2]
```diff
@@ -254,16 +292,24 @@ CS_VARIABLE* find_var_from_pools(CSOUND* csound, char* varName, char* varBaseNam
...
                                        varBaseName);
   } else {
+    // now we check for local variables
     var = csoundFindVariableWithName(csound, typeTable->localPool,
                                      varBaseName);
+    // then check for global variables that may have been explicitly defined
+    if(var == NULL) 
+    var = csoundFindVariableWithName(csound, csound->engineState.varPool,
+                                     varBaseName);
+    if(var == NULL)
```
Comment (kunstmusik) [inline Engine/csound_orc_semantics.c:311]: Seems like we already checked the global pool earlier? (line 299)

**Comment by kunstmusik** [comment]
Went further through the code changes and saw that there was indeed a scoping system added to check local then global pools. I still wonder if we couldn't do without @global annotation and just use this to get explicitly typed variables in instr0 to work as global vars.

**Comment by vlazzarini** [comment]
This would break compatibility since we can have local variables in instr0 (all the ones without g), so we need to make a distinction between them. 

I commonly use local variables (counters etc) in instr0, which have the same names as ones instruments. So we could not export them by default. The use of g ensures that. I have to check, but I think instr 0 local vars are not persistent from compilation to compilation.

Having that rule from implicit types makes it hard for us to have a different rule for explicit without it being confusing. We need an annotation as we have in 'g' for implicit types.

I don't think we can avoid an annotation: if I have

icount init 0 ; var in instr0

instr 1
icount init 0 ; different var in instr0
endin

it needs to work the same as

count:i  init 0

instr 1
count:i init 0
endin

and without an annotation or an export symbol/opcode/function we can't make a distinction.

**Comment by vlazzarini** [comment]
Also: we can declare global variables inside instruments. We would have a different rule for explicit variables where we cannot,  that to me is also confusing.

File: Engine/csound_orc_semantics.c (hunk 2 above)
Comment (vlazzarini) [reply Engine/csound_orc_semantics.c:311]: only for certain implicit types and reserved words, so we need to check again.

Comment (vlazzarini) [reply Engine/csound_orc_semantics.c:311]: only for certain implicit type vars and reserved words, so we need to check again.

File: Engine/csound_orc.lex (hunk 1 above)
Comment (vlazzarini) [reply Engine/csound_orc.lex]: you are right, improved.

**Comment by vlazzarini** [comment]
Furthermore, consider you have this code

```
ivar = 440
instr 1
ivar = p4
a1 oscili 0.1,ivar
out a1
endin
schedule(1, 0, 1, ivar)
schedule(1, 0, 1, ivar*1.5)
```

and you learned about explicit-type vars and want to modernise your
code, 

```
var:i = 440
instr 1
var:i = p4
a1 oscili 0.1, var
out a1
endin
schedule(1, 0, 1, var)
schedule(1, 0, 1, var*1.5)
```

then you get all confused as it doesn't work as before.

**Comment by kunstmusik** [comment]
I see, I suppose in JS they got around this once they introduce let and const to declare variables locally. If we did something similar (not saying we should), it'd look like:

```
var:i = 440
instr 1
  let var:i = p4
  a1 oscili 0.1, var
  out a1
endin
schedule(1, 0, 1, var)
schedule(1, 0, 1, var*1.5)
```

to differentiate the global vs locally scoped vars. 

Python goes a different route and needs to have global added to reference the global var:

```
var = 3
def x():
  global var
  print var
```

and we'd have something like:

```
var:i = 440
instr 1
  global var
  var2:i = p4
  a1 = oscili(0.1, var)
  a1 += oscili(0.1, var2)
  out a1
endin
schedule(1, 0, 1, var)
schedule(1, 0, 1, var*1.5)
```

Interesting to think about the different approaches. I suppose it's strange that we can define global variables within instruments but we're stuck with it due to legacy implementation. (Probably not a good practice!)

One concern that popped up in my head was what would all of this look like if/when we add namespaces and want to reference a namespaced-scoped variable.

**Comment by vlazzarini** [comment]
I proposed that kind of approach in the open issue discussing this, but people there seem not to like it because it was different from how we use "g" and they thought it was too complicated. They supported the use of annotations, so I went with it.

**Comment by rorywalsh** [comment]
I think the fact that we don't need to introduce any new scoping opcodes to our instruments in one big advantage of using annotations. @joachimheintz do you have any thought on this?

**Comment by joachimheintz** [comment]
i agree!

**Comment by rorywalsh** [comment]
@vlazzarini @kunstmusik will the use of annotations mess up future namespace features? I guess if it does this needs to be revisited, and if not, we can maybe push through and accept this.

**Comment by vlazzarini** [comment]
Just thinking out loud, since we have g-vars would we not end up with issues when introducing namespaces (if any arise) which would be the same in case we introduce annotations for explicit vars?

All we doing is rolling out an already existing mechanism to work with explicit vars.

**Comment by kunstmusik** [comment]
I think the way I'm seeing it is if we have "@global" then it would suggest we do something similarly to access namespaced data since we're doing a similar operation of accessing data from a different scope. We wouldn't have to and could always have "@global" and another system for accessing a namespaced value, but want to make sure we don't design ourselves into a hole.

**Comment by vlazzarini** [comment]
I would think we could accommodate a suitable namespace annotation with this scheme. Notice that @global only matters at variable declaration, not access.

**Comment by giovannibedetti** [comment]
why not:

var:gi = 440

it seems very clear but I'm not sure how feasable it will be with the parser

**Comment by kunstmusik** [comment]
@giovannibedetti globalness is part of the variable, not the datatype. (That's how it's in my head at least.)

**Comment by vlazzarini** [comment]
that's how I understand it too. Also it wouldn't look great with the new types like `var:gInstrDef` or `var:gComplex`

**Comment by vlazzarini** [comment]
An alternative is to add the annotation at the front,```global:var:i```, but it may complicate the parsing.

File: Engine/csound_orc_compile.c [hunk 3]
```diff
@@ -839,7 +779,8 @@ INSTRTXT *create_instrument0(CSOUND *csound, TREE *root,
 
   if (csound->ksmps != DFLT_KSMPS) {
     reallocateVarPoolMemory(csound, engineState->varPool);
-    // csound->Message(csound, "recalculate varpool\n");
+    if(csound->GetDebug(csound))
```
Comment (kunstmusik) [inline Engine/csound_orc_compile.c:782]: Sidenote: this seems like a common pattern, we might benefit from a macro or function such as `debugMessage(csound, "some string")`

File: Engine/csound_orc_compile.c [hunk 4]
```diff
@@ -1455,15 +1381,16 @@ void insert_opcodes(CSOUND *csound, OPCODINFO *opcodeInfo,
...
+        engineState->instrtxtp = (INSTRTXT **)
+          csound->ReAlloc(csound, engineState->instrtxtp,
+                          (1 + engineState->maxopcno) * sizeof(INSTRTXT *));
         /* Array expected to be nulled so.... */
         while (++i <= engineState->maxopcno)
           engineState->instrtxtp[i] = NULL;
       }
       inm->instno = num;
-      // csound->Message(csound, Str("UDO INSTR NUM: %d\n"), num);
+      if(csound->GetDebug(csound))
```
Comment (kunstmusik) [inline Engine/csound_orc_compile.c:1392]: Related to earlier comment, we might find it better to introduce a csound->Log(csound, log_level, formatted string, args...) method that introduces different log levels like info, warn, error, etc. This is pretty common and would hide things like checking GetDebug() and simplify the codebase IMO.

**Comment by kunstmusik** [comment]
@vlazzarini Suggest we merge but consider this a preview feature and continue discussion and experimentation, perhaps following up with a later PR if we decide on a different path. I think there's some fixes for other things in this branch that would be to get in now. Sound good to you?

**Comment by vlazzarini** [comment]
sounds good, that'll allow me to do tests with global vars for implicit types and if we decide to re-implement this another, we can just replace the mechanism.

**Review by kunstmusik — APPROVED** [review]

**Pull request merged**